File: packages/earthsci_toolkit/src/earthsci_toolkit/cli/pde_simulation_adapter.py
"""Python adapter for the cross-language PDE-simulation conformance tier (ess-fmw).

Drives the shared, pre-discretized method-of-lines fixtures listed in
``tests/conformance/pde_simulation/manifest.json``. For every fixture it:

  * evaluates the discretized RHS f(u, t) at each declared probe state via
    :func:`earthsci_toolkit.evaluate_rhs` (the NumPy-interpreter RHS, no
    integrator), and
  * integrates the trajectory from the declared initial conditions with the
    pinned integrator (SciPy ``solve_ivp`` + the manifest's method/rtol/atol),
    sampling at the declared output times.

The runner discovers it via ``$EARTHSCI_PDE_SIM_ADAPTER_PYTHON`` or on PATH:

    earthsci-pde-sim-adapter-python --manifest <manifest.json> --output <out.json>

Emits ``{"binding":"python","fixtures":{<id>:{"rhs":{<probe>:{name:val}},
"trajectory":{<tstr>:{name:val}}}}}`` with bare ``u[i]`` element names.
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any, Dict

import numpy as np

import earthsci_toolkit as et
from earthsci_toolkit import evaluate_rhs, simulate


def _bare(name: str) -> str:
    return name.split(".", 1)[1] if "." in name else name


def _time_key(t: float) -> str:
    return f"{float(t):g}"


def _sample_trajectory(result, out_times) -> Dict[str, Dict[str, float]]:
    """Interpolate every state element of a SimulationResult at each output
    time, keyed by bare element name."""
    traj: Dict[str, Dict[str, float]] = {}
    for t in out_times:
        col: Dict[str, float] = {}
        for row, name in enumerate(result.vars):
            col[_bare(name)] = float(np.interp(float(t), result.t, result.y[row]))
        traj[_time_key(t)] = col
    return traj


def run_fixture(fixture: dict, base: Path, integ: dict) -> Dict[str, Any]:
    esm = et.load(str(base / fixture["path"]))

    rhs: Dict[str, Dict[str, float]] = {}
    for probe in fixture["rhs_probes"]:
        raw = evaluate_rhs(esm, dict(probe["state"]), t=float(probe.get("t", 0.0)))
        rhs[probe["id"]] = {_bare(k): float(v) for k, v in raw.items()}

    tr = fixture["trajectory"]
    tspan = (float(tr["time_span"]["start"]), float(tr["time_span"]["end"]))
    result = simulate(
        esm, tspan,
        initial_conditions=dict(tr["initial_conditions"]),
        method=integ.get("method", "RK45"),
        rtol=float(integ.get("rtol", 1e-10)),
        atol=float(integ.get("atol", 1e-12)),
    )
    traj = _sample_trajectory(result, tr["output_times"])
    return {"rhs": rhs, "trajectory": traj}


def main(argv=None) -> int:
    parser = argparse.ArgumentParser(description="Python PDE-simulation conformance adapter")
    parser.add_argument("--manifest", required=True, type=Path)
    parser.add_argument("--output", required=True, type=Path)
    args = parser.parse_args(argv)

    manifest = json.loads(args.manifest.read_text())
    integ = manifest.get("integrators", {}).get("python", {})
    base = args.manifest.parent

    fixtures: Dict[str, Any] = {}
    for fixture in manifest["fixtures"]:
        try:
            fixtures[fixture["id"]] = run_fixture(fixture, base, integ)
        except Exception as exc:  # noqa: BLE001 - surface per-fixture failure to the runner
            fixtures[fixture["id"]] = {"error": f"{type(exc).__name__}: {exc}"}

    payload = {"binding": "python", "fixtures": fixtures}
    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(json.dumps(payload, indent=2, sort_keys=True) + "\n")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: packages/EarthSciSerialization.jl/scripts/pde_simulation_adapter.jl
# Julia adapter for the cross-language PDE-simulation conformance tier (ess-fmw).
#
# Reference binding. For every fixture in the manifest it:
#   * evaluates the discretized RHS f(u, t) at each declared probe state via the
#     MTK-free tree-walk evaluator (`build_evaluator` -> `f!(du, u, p, t)`), and
#   * integrates the trajectory from the declared initial conditions with the
#     pinned integrator (Tsit5 + manifest reltol/abstol), sampling at the
#     declared output times.
#
# Output element names are the bare `u[i]` / `u[i,j]` slot names from the
# evaluator's var map (shared with Python/Rust). Emits, to --output:
#   {"binding":"julia","fixtures":{<id>:{"rhs":{<probe>:{name:val}},
#                                        "trajectory":{<tstr>:{name:val}}}}}
#
# Invoked with the dedicated env that carries OrdinaryDiffEqTsit5 + JSON3:
#   julia --project=packages/EarthSciSerialization.jl/scripts/pde_sim_adapter \
#         packages/EarthSciSerialization.jl/scripts/pde_simulation_adapter.jl \
#         --manifest <manifest.json> --output <out.json>

# Self-contained environment bootstrap. The dedicated adapter project
# (scripts/pde_sim_adapter/Project.toml) pins EarthSciSerialization (dev'd from
# ../..) + OrdinaryDiffEqTsit5 + JSON3. Manifest.toml is gitignored repo-wide,
# so on a fresh checkout we re-establish the local dev path then instantiate; on
# warm runs (Manifest already present) this is just a fast resolve check.
import Pkg
let env = joinpath(@__DIR__, "pde_sim_adapter")
    Pkg.activate(env; io=devnull)
    isfile(joinpath(env, "Manifest.toml")) ||
        Pkg.develop(path=normpath(joinpath(@__DIR__, "..")); io=devnull)
    Pkg.instantiate(; io=devnull)
end

using EarthSciSerialization
using JSON3
import OrdinaryDiffEqTsit5
const ODE = OrdinaryDiffEqTsit5

function parse_args(args)
    manifest = nothing
    output = nothing
    i = 1
    while i <= length(args)
        if args[i] == "--manifest"
            manifest = args[i + 1]; i += 2
        elseif args[i] == "--output"
            output = args[i + 1]; i += 2
        else
            i += 1
        end
    end
    manifest === nothing && error("--manifest is required")
    output === nothing && error("--output is required")
    (manifest, output)
end

# Trajectory time key: a plain float string. The Python harness re-normalizes
# every key via float(k):g, so the exact rendering here only has to round-trip.
tkey(t) = string(float(t))

_ic_dict(obj) = Dict{String,Float64}(String(k) => Float64(v) for (k, v) in pairs(obj))

function rhs_at(model, probe_state, t)
    ics = _ic_dict(probe_state)
    f!, u0, p, _, vmap = build_evaluator(model; initial_conditions=ics)
    du = similar(u0)
    f!(du, u0, p, Float64(t))
    Dict{String,Float64}(name => Float64(du[idx]) for (name, idx) in vmap)
end

function trajectory(model, ic, t0, t1, out_times, reltol, abstol)
    ics = _ic_dict(ic)
    f!, u0, p, _, vmap = build_evaluator(model; initial_conditions=ics)
    prob = ODE.ODEProblem(f!, u0, (Float64(t0), Float64(t1)), p)
    sol = ODE.solve(prob, ODE.Tsit5(); reltol=reltol, abstol=abstol)
    out = Dict{String,Any}()
    for t in out_times
        st = sol(Float64(t))
        out[tkey(t)] = Dict{String,Float64}(name => Float64(st[idx])
                                            for (name, idx) in vmap)
    end
    out
end

function main()
    manifest_path, output_path = parse_args(ARGS)
    manifest = JSON3.read(read(manifest_path, String))
    integ = manifest.integrators.julia
    reltol = Float64(integ.reltol)
    abstol = Float64(integ.abstol)
    base = dirname(manifest_path)

    fixtures = Dict{String,Any}()
    for fx in manifest.fixtures
        path = joinpath(base, String(fx.path))
        file = load(path)
        model = file.models[String(fx.model)]

        rhs = Dict{String,Any}()
        for pr in fx.rhs_probes
            rhs[String(pr.id)] = rhs_at(model, pr.state, pr.t)
        end

        tr = fx.trajectory
        ts = tr.time_span
        traj = trajectory(model, tr.initial_conditions,
                          ts[Symbol("start")], ts[Symbol("end")],
                          tr.output_times, reltol, abstol)

        fixtures[String(fx.id)] = Dict("rhs" => rhs, "trajectory" => traj)
    end

    payload = Dict("binding" => "julia", "fixtures" => fixtures)
    open(output_path, "w") do io
        JSON3.write(io, payload)
    end
end

main()
